clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -3583,24 +3585,44 @@ void populateCIRToLLVMConversionPatterns(mlir::RewritePatternSet &patterns,
 }
 
 namespace {
+
+using IRASMap = mlir::cir::AddressSpaceAttr::MapTy;
+
+/// Returns the CIR address space map when available. Returns nullptr if the
+/// triple is not present in the module.
+IRASMap const *getIRAddrSpaceMap(mlir::ModuleOp module) {
+  // If the triple is not present, e.g. CIR modules parsed from text, we
+  // cannot init LowerModule properly.
+  assert(!::cir::MissingFeatures::makeTripleAlwaysPresent());
+  // Here we will use a default AS map to ignore all AS stuff.
+  if (!module->hasAttr("cir.triple"))
+    return nullptr;
```

Comment:
Note that `IRASMap const *` is a pointer to array. It would be `nullptr` if triple is not present.

If no address space attributes are present in CIR, we ensure that a null `IRASMap` remains functional. However, if there is an address space attribute to convert, the guard at line 3619 will reject it.

Reply:
I thought a bit about using the pointer here, but couldn't come up with a better design. I believe using `std::optional<std::reference_wrapper<IRASMap>>` would work, but that doesn't really simplify anything.

The underlying question is, do we really need this miscompilation protection here or could we just assume a default mapping similar to always having a `DataLayout`? In case a target triple is present and the target doesn't supply an AS map (or an incomplete AS map), we'd happily use a default flattening to (target) AS 0. To catch that, we'd need to use an actual map (e.g. `SmallDenseMap`) to let targets opt-in to the CIR AS they support, but that design point would deviate from OG clang.

Reply:
To clarify, for this PR I'm suggesting to either (1) keep the "no triple -> null map" logic or (2) drop the check and use a default map when no triple is present. I meant (3) "targets opt-in to supported AS" more as food for thought and potential future improvement that should not hold up this PR which finally reactivates the in-tree OpenCL tests 🎉